diff --git a/edifyscripting.c b/edifyscripting.c
index e66df2b..5cd0952 100644
--- a/edifyscripting.c
+++ b/edifyscripting.c
@@ -185,8 +185,11 @@ Value* RestoreFn(const char* name, State* state, int argc, Expr* argv[]) {
     
     char* path = strdup(args2[0]);
     int restoreboot = 1;
-    int restoresystem = 1;
-    int restoredata = 1;
+    int restoresystem1 = 1;
+    int restoredata1 = 1;
+    int restoresystem2 = 1;
+    int restoredatadata = 1;
+    int restoredata2 = 1;
     int restorecache = 1;
     int restoresdext = 1;
     int i;
@@ -196,10 +199,16 @@ Value* RestoreFn(const char* name, State* state, int argc, Expr* argv[]) {
             continue;
         if (strcmp(args2[i], "noboot") == 0)
             restoreboot = 0;
-        else if (strcmp(args2[i], "nosystem") == 0)
-            restoresystem = 0;
-        else if (strcmp(args2[i], "nodata") == 0)
-            restoredata = 0;
+        else if (strcmp(args2[i], "nosystem1") == 0)
+            restoresystem1 = 0;
+        else if (strcmp(args2[i], "nodata1") == 0)
+            restoredata1 = 0;
+	else if (strcmp(args2[i], "nosystem2") == 0)
+            restoresystem2 = 0;
+        else if (strcmp(args2[i], "nodata2") == 0)
+            restoredata2 = 0;
+        else if (strcmp(args2[i], "nodatadata") == 0)
+            restoredatadata = 0;
         else if (strcmp(args2[i], "nocache") == 0)
             restorecache = 0;
         else if (strcmp(args2[i], "nosd-ext") == 0)
@@ -212,7 +221,7 @@ Value* RestoreFn(const char* name, State* state, int argc, Expr* argv[]) {
     free(args);
     free(args2);
 
-    if (0 != nandroid_restore(path, restoreboot, restoresystem, restoredata, restorecache, restoresdext, 0)) {
+    if (0 != nandroid_restore(path, restoreboot, restoresystem1, restoredata1, restoresystem2, restoredata2, restoredatadata, restorecache, restoresdext, 0)) {
         free(path);
         return StringValue(strdup(""));
     }
diff --git a/etc/init.rc b/etc/init.rc
index 832b864..c4fac45 100644
--- a/etc/init.rc
+++ b/etc/init.rc
@@ -14,8 +14,10 @@ on init
     mkdir /sd-ext
     mkdir /datadata
     mkdir /emmc
-    mkdir /system
-    mkdir /data
+    mkdir /system1
+    mkdir /data1
+    mkdir /system2
+    mkdir /data2
     mkdir /cache
     mount /tmp /tmp tmpfs
 
diff --git a/extendedcommands.c b/extendedcommands.c
index 845626a..99646bc 100644
--- a/extendedcommands.c
+++ b/extendedcommands.c
@@ -354,7 +354,7 @@ void show_nandroid_restore_menu(const char* path)
         return;
 
     if (confirm_selection("Confirm restore?", "Yes - Restore"))
-        nandroid_restore(file, 1, 1, 1, 1, 1, 0);
+        nandroid_restore(file, 1, 1, 1, 1, 1, 1, 1, 1, 0);
 }
 
 #ifndef BOARD_UMS_LUNFILE
@@ -773,8 +773,11 @@ void show_nandroid_advanced_restore_menu(const char* path)
     };
 
     static char* list[] = { "Restore boot",
-                            "Restore system",
-                            "Restore data",
+                            "Restore system1",
+                            "Restore data1",
+                            "Restore system2",
+                            "Restore data2",
+                            "Restore datadata",
                             "Restore cache",
                             "Restore sd-ext",
                             "Restore wimax",
@@ -793,27 +796,39 @@ void show_nandroid_advanced_restore_menu(const char* path)
     {
         case 0:
             if (confirm_selection(confirm_restore, "Yes - Restore boot"))
-                nandroid_restore(file, 1, 0, 0, 0, 0, 0);
+                nandroid_restore(file, 1, 0, 0, 0, 0, 0, 0, 0, 0);
             break;
         case 1:
-            if (confirm_selection(confirm_restore, "Yes - Restore system"))
-                nandroid_restore(file, 0, 1, 0, 0, 0, 0);
+            if (confirm_selection(confirm_restore, "Yes - Restore system1"))
+                nandroid_restore(file, 0, 1, 0, 0, 0, 0, 0, 0, 0);
             break;
         case 2:
-            if (confirm_selection(confirm_restore, "Yes - Restore data"))
-                nandroid_restore(file, 0, 0, 1, 0, 0, 0);
+            if (confirm_selection(confirm_restore, "Yes - Restore data1"))
+                nandroid_restore(file, 0, 0, 1, 0, 0, 0, 0, 0, 0);
             break;
         case 3:
-            if (confirm_selection(confirm_restore, "Yes - Restore cache"))
-                nandroid_restore(file, 0, 0, 0, 1, 0, 0);
+            if (confirm_selection(confirm_restore, "Yes - Restore system2"))
+                nandroid_restore(file, 0, 0, 0, 1, 0, 0, 0, 0, 0);
             break;
         case 4:
-            if (confirm_selection(confirm_restore, "Yes - Restore sd-ext"))
-                nandroid_restore(file, 0, 0, 0, 0, 1, 0);
+            if (confirm_selection(confirm_restore, "Yes - Restore data2"))
+                nandroid_restore(file, 0, 0, 0, 0, 1, 0, 0, 0, 0);
             break;
         case 5:
+            if (confirm_selection(confirm_restore, "Yes - Restore datadata"))
+                nandroid_restore(file, 0, 0, 0, 0, 0, 1, 0, 0, 0);
+            break;
+        case 6:
+            if (confirm_selection(confirm_restore, "Yes - Restore cache"))
+                nandroid_restore(file, 0, 0, 0, 0, 0, 0, 1, 0, 0);
+            break;
+        case 7:
+            if (confirm_selection(confirm_restore, "Yes - Restore sd-ext"))
+                nandroid_restore(file, 0, 0, 0, 0, 0, 0, 0, 1, 0);
+            break;
+        case 8:
             if (confirm_selection(confirm_restore, "Yes - Restore wimax"))
-                nandroid_restore(file, 0, 0, 0, 0, 0, 1);
+                nandroid_restore(file, 0, 0, 0, 0, 0, 0, 0, 0, 1);
             break;
     }
 }
@@ -893,9 +908,11 @@ void show_nandroid_menu()
 
 void wipe_battery_stats()
 {
-    ensure_path_mounted("/data");
+    ensure_path_mounted("/data1");
+    ensure_path_mounted("/data2");
     remove("/data/system/batterystats.bin");
-    ensure_path_unmounted("/data");
+    ensure_path_unmounted("/data1");
+    ensure_path_unmounted("/data2");
     ui_print("Battery Stats wiped.\n");
 }
 
@@ -907,7 +924,8 @@ void show_advanced_menu()
     };
 
     static char* list[] = { "Reboot Recovery",
-                            "Wipe Dalvik Cache",
+                            "Wipe System1 Dalvik Cache",
+                            "Wipe System2 Dalvik Cache",
                             "Wipe Battery Stats",
                             "Report Error",
                             "Key Test",
@@ -934,29 +952,44 @@ void show_advanced_menu()
             }
             case 1:
             {
-                if (0 != ensure_path_mounted("/data"))
+                if (0 != ensure_path_mounted("/data1"))
                     break;
                 ensure_path_mounted("/sd-ext");
                 ensure_path_mounted("/cache");
-                if (confirm_selection( "Confirm wipe?", "Yes - Wipe Dalvik Cache")) {
-                    __system("rm -r /data/dalvik-cache");
+                if (confirm_selection( "Confirm wipe?", "Yes - Wipe System1 Dalvik Cache")) {
+                    __system("rm -r /data1/dalvik-cache");
                     __system("rm -r /cache/dalvik-cache");
                     __system("rm -r /sd-ext/dalvik-cache");
                     ui_print("Dalvik Cache wiped.\n");
                 }
-                ensure_path_unmounted("/data");
+                ensure_path_unmounted("/data1");
                 break;
             }
             case 2:
             {
+                if (0 != ensure_path_mounted("/data2"))
+                    break;
+                ensure_path_mounted("/sd-ext");
+                ensure_path_mounted("/cache");
+                if (confirm_selection( "Confirm wipe?", "Yes - Wipe System2 Dalvik Cache")) {
+                    __system("rm -r /data2/dalvik-cache");
+                    __system("rm -r /cache/dalvik-cache");
+                    __system("rm -r /sd-ext/dalvik-cache");
+                    ui_print("Dalvik Cache wiped.\n");
+                }
+                ensure_path_unmounted("/data2");
+                break;
+            }
+            case 3:
+            {
                 if (confirm_selection( "Confirm wipe?", "Yes - Wipe Battery Stats"))
                     wipe_battery_stats();
                 break;
             }
-            case 3:
+            case 4:
                 handle_failure(1);
                 break;
-            case 4:
+            case 5:
             {
                 ui_print("Outputting key codes.\n");
                 ui_print("Go back to end debugging.\n");
@@ -971,12 +1004,12 @@ void show_advanced_menu()
                 while (action != GO_BACK);
                 break;
             }
-            case 5:
+            case 6:
             {
                 ui_printlogtail(12);
                 break;
             }
-            case 6:
+            case 7:
             {
                 static char* ext_sizes[] = { "128M",
                                              "256M",
@@ -1019,7 +1052,7 @@ void show_advanced_menu()
                     ui_print("An error occured while partitioning your SD Card. Please see /tmp/recovery.log for more details.\n");
                 break;
             }
-            case 7:
+            case 8:
             {
                 ensure_path_mounted("/system");
                 ensure_path_mounted("/data");
@@ -1028,7 +1061,7 @@ void show_advanced_menu()
                 ui_print("Done!\n");
                 break;
             }
-            case 8:
+            case 9:
             {
                 static char* ext_sizes[] = { "128M",
                                              "256M",
@@ -1108,10 +1141,12 @@ void create_fstab()
     if (NULL != vol && strcmp(vol->fs_type, "mtd") != 0 && strcmp(vol->fs_type, "emmc") != 0 && strcmp(vol->fs_type, "bml") != 0)
          write_fstab_root("/boot", file);
     write_fstab_root("/cache", file);
-    write_fstab_root("/data", file);
+    write_fstab_root("/data1", file);
+    write_fstab_root("/data2", file);
     write_fstab_root("/datadata", file);
     write_fstab_root("/emmc", file);
-    write_fstab_root("/system", file);
+    write_fstab_root("/system1", file);
+    write_fstab_root("/system2", file);
     write_fstab_root("/sdcard", file);
     write_fstab_root("/sd-ext", file);
     fclose(file);
@@ -1155,8 +1190,10 @@ void process_volumes() {
 
     ui_print("Checking for ext4 partitions...\n");
     int ret = 0;
-    ret = bml_check_volume("/system");
-    ret |= bml_check_volume("/data");
+    ret = bml_check_volume("/system1");
+    ret |= bml_check_volume("/data1");
+    ret = bml_check_volume("/system2");
+    ret |= bml_check_volume("/data2");
     if (has_datadata())
         ret |= bml_check_volume("/datadata");
     ret |= bml_check_volume("/cache");
@@ -1182,7 +1219,7 @@ void process_volumes() {
     ui_print("in case of error.\n");
 
     nandroid_backup(backup_path);
-    nandroid_restore(backup_path, 1, 1, 1, 1, 1, 0);
+    nandroid_restore(backup_path, 1, 1, 1, 1, 1, 1, 1, 1, 0);
     ui_set_show_text(0);
 }
 
diff --git a/nandroid.c b/nandroid.c
index fc39359..00350d7 100644
--- a/nandroid.c
+++ b/nandroid.c
@@ -275,12 +275,19 @@ int nandroid_backup(const char* backup_path)
             return print_and_error("Error while dumping WiMAX image!\n");
     }
 
-    if (0 != (ret = nandroid_backup_partition(backup_path, "/system")))
+    if (0 != (ret = nandroid_backup_partition(backup_path, "/system1")))
         return ret;
 
-    if (0 != (ret = nandroid_backup_partition(backup_path, "/data")))
+    if (0 != (ret = nandroid_backup_partition(backup_path, "/data1")))
         return ret;
 
+    if (0 != (ret = nandroid_backup_partition(backup_path, "/system2")))
+        return ret;
+
+    if (0 != (ret = nandroid_backup_partition(backup_path, "/data2")))
+        return ret;
+
+
     if (has_datadata()) {
         if (0 != (ret = nandroid_backup_partition(backup_path, "/datadata")))
             return ret;
@@ -523,7 +530,7 @@ int nandroid_restore_partition(const char* backup_path, const char* root) {
     return nandroid_restore_partition_extended(backup_path, root, 1);
 }
 
-int nandroid_restore(const char* backup_path, int restore_boot, int restore_system, int restore_data, int restore_cache, int restore_sdext, int restore_wimax)
+int nandroid_restore(const char* backup_path, int restore_boot, int restore_system1, int restore_data1,int restore_system2, int restore_data2, int restore_data, int restore_cache, int restore_sdext, int restore_wimax)
 {
     ui_set_background(BACKGROUND_ICON_INSTALLING);
     ui_show_indeterminate_progress();
@@ -573,10 +580,16 @@ int nandroid_restore(const char* backup_path, int restore_boot, int restore_syst
         }
     }
 
-    if (restore_system && 0 != (ret = nandroid_restore_partition(backup_path, "/system")))
+    if (restore_system1 && 0 != (ret = nandroid_restore_partition(backup_path, "/system1")))
+        return ret;
+
+    if (restore_data1 && 0 != (ret = nandroid_restore_partition(backup_path, "/data1")))
+        return ret;   
+   
+    if (restore_system2 && 0 != (ret = nandroid_restore_partition(backup_path, "/system2")))
         return ret;
 
-    if (restore_data && 0 != (ret = nandroid_restore_partition(backup_path, "/data")))
+    if (restore_data2 && 0 != (ret = nandroid_restore_partition(backup_path, "/data2")))
         return ret;
         
     if (has_datadata()) {
@@ -626,7 +639,7 @@ int nandroid_main(int argc, char** argv)
     {
         if (argc != 3)
             return nandroid_usage();
-        return nandroid_restore(argv[2], 1, 1, 1, 1, 1, 0);
+        return nandroid_restore(argv[2], 1, 1, 1, 1, 1, 1, 1, 1, 0);
     }
     
     return nandroid_usage();
diff --git a/nandroid.h b/nandroid.h
index 836f192..5ec26e7 100644
--- a/nandroid.h
+++ b/nandroid.h
@@ -3,6 +3,6 @@
 
 int nandroid_main(int argc, char** argv);
 int nandroid_backup(const char* backup_path);
-int nandroid_restore(const char* backup_path, int restore_boot, int restore_system, int restore_data, int restore_cache, int restore_sdext, int restore_wimax);
+int nandroid_restore(const char* backup_path, int restore_boot, int restore_system1, int restore_data1, int restore_system2, int restore_data2, int restore_data, int restore_cache, int restore_sdext, int restore_wimax);
 
-#endif
\ No newline at end of file
+#endif
